File: tools/emit_function.py
#!/usr/bin/env python3
import argparse
import re
import subprocess
from pathlib import Path

from discover import Discovery, ROM_BASE, sx


def branch_target(data, address, kind):
    offset = address - ROM_BASE
    half = int.from_bytes(data[offset:offset + 2], "little")
    if kind == "branch":
        return address + 4 + (sx(half & 0x7FF, 11) << 1)
    if kind == "conditional":
        return address + 4 + (sx(half & 0xFF, 8) << 1)
    return None


def disassemble(data, start, end, output_dir):
    binary = output_dir / f"{start:08x}.bin"
    binary.write_bytes(data[start - ROM_BASE:end - ROM_BASE])
    text = subprocess.check_output([
        "arm-none-eabi-objdump", "-D", "-b", "binary", "-marm",
        "-Mforce-thumb", f"--adjust-vma=0x{start:08x}", str(binary),
    ], text=True)
    result = {}
    pattern = re.compile(
        r"^\s*([0-9a-f]+):\s+(?:[0-9a-f]{4}\s+)+(.*)$")
    for line in text.splitlines():
        match = pattern.match(line)
        if match:
            result[int(match.group(1), 16)] = match.group(2).strip()
    return result


def emit_discovery(rom, discovery, address, output):
    function = discovery.functions.get(address)
    if function is None:
        raise ValueError(f"unknown function: {address:08x}")
    if function["mode"] != "thumb":
        raise ValueError("only Thumb functions are currently supported")
    instructions = sorted(function["instructions"])
    start = instructions[0]
    end = max(
        item + discovery.instructions[item]["size"] for item in instructions)
    output_dir = output.parent
    output_dir.mkdir(parents=True, exist_ok=True)
    decoded = disassemble(rom, start, end, output_dir)
    calls = {source: target for source, target, mode in discovery.calls
             if source in function["instructions"] and mode == "thumb"}
    labels = {
        target for item in instructions
        for target in [branch_target(
            rom, item, discovery.instructions[item]["kind"])]
        if target in function["instructions"]
    }
    literals = {}
    lines = [".syntax unified", ".thumb", f"glabel Func_{address:08x}"]
    for item in instructions:
        if item in labels:
            lines.append(f".L_{item:08x}:")
        instruction = decoded[item].split(";", 1)[0].strip()
        kind = discovery.instructions[item]["kind"]
        if kind == "call" and item in calls:
            mnemonic = instruction.split(None, 1)[0]
            instruction = f"{mnemonic} Func_{calls[item]:08x}"
        elif kind in ("branch", "conditional"):
            target = branch_target(rom, item, kind)
            if target in function["instructions"]:
                mnemonic = instruction.split(None, 1)[0]
                instruction = f"{mnemonic} .L_{target:08x}"
        half = discovery.u16(item)
        if half & 0xF800 == 0x4800:
            register = (half >> 8) & 7
            literal = ((item + 4) & ~3) + ((half & 0xFF) << 2)
            literals[literal] = discovery.u32(literal)
            instruction = f"ldr r{register}, .L_{literal:08x}"
        lines.append(f"    {instruction}")
    for literal, value in sorted(literals.items()):
        lines.extend([
            "    .align 2",
            f".L_{literal:08x}:",
            f"    .word 0x{value:08x}",
        ])
    output.write_text("\n".join(lines) + "\n")


def emit(rom, address, output):
    discovery = Discovery(rom)
    discovery.run()
    emit_discovery(rom, discovery, address, output)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("rom", type=Path)
    parser.add_argument("address", type=lambda value: int(value, 0))
    parser.add_argument("-o", "--output", type=Path, required=True)
    args = parser.parse_args()
    emit(args.rom.read_bytes(), args.address & ~1, args.output)


if __name__ == "__main__":
    main()

File: tools/emit_function.ts
// Tool role: library; imported by tools/decompile_batch.ts, tools/overlay_match.ts.
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { Discovery, ROM_BASE, sx, type Instruction } from "./discover.ts";

function hex(value: number): string {
  return value.toString(16).padStart(8, "0");
}

export function branchTarget(data: Uint8Array, address: number, kind: string, base = ROM_BASE): number | null {
  const offset = address - base;
  const half = Buffer.from(data).readUInt16LE(offset);
  if (kind === "branch") return address + 4 + (sx(half & 0x7ff, 11) << 1);
  if (kind === "conditional") return address + 4 + (sx(half & 0xff, 8) << 1);
  return null;
}
export const branch_target = branchTarget;

export function disassemble(
  data: Uint8Array,
  start: number,
  end: number,
  instructions: Map<number, Instruction>,
  outputDir: string,
  base = ROM_BASE,
): Map<number, string> {
  const binary = join(outputDir, `${hex(start)}.bin`);
  const masked = Buffer.alloc(Math.ceil((end - start) / 2) * 2);
  for (let offset = 0; offset < masked.length; offset += 2) {
    masked[offset] = 0xc0;
    masked[offset + 1] = 0x46;
  }
  const sourceData = Buffer.from(data);
  for (const [address, instruction] of instructions) {
    const source = address - base;
    sourceData.copy(masked, address - start, source, source + instruction.size);
  }
  writeFileSync(binary, masked.subarray(0, end - start));
  const command = ["arm-none-eabi-objdump", "-D", "-b", "binary", "-marm", "-Mforce-thumb", `--adjust-vma=0x${hex(start)}`, binary];
  const process = Bun.spawnSync(command, { stdout: "pipe", stderr: "pipe" });
  if (process.exitCode !== 0) throw new Error(process.stderr.toString().trim());
  const result = new Map<number, string>();
  const pattern = /^\s*([0-9a-f]+):\s+(?:[0-9a-f]{4}\s+)+(.*)$/;
  for (const line of process.stdout.toString().split(/\r?\n/)) {
    const match = pattern.exec(line);
    if (match) result.set(Number.parseInt(match[1], 16), match[2].trim());
  }
  return result;
}

export function emitDiscovery(rom: Uint8Array, discovery: Discovery, address: number, output: string): void {
  const fn = discovery.functions.get(address);
  if (!fn) throw new Error(`unknown function: ${hex(address)}`);
  if (fn.mode !== "thumb") throw new Error("only Thumb functions are currently supported");
  const instructions = [...fn.instructions].sort((a, b) => a - b);
  const start = instructions[0];
  const end = Math.max(...instructions.map((item) => item + discovery.instructions.get(item)!.size));
  const outputDir = dirname(output);
  mkdirSync(outputDir, { recursive: true });
  const functionInstructions = new Map(instructions.map((item) => [item, discovery.instructions.get(item)!]));
  const decoded = disassemble(rom, start, end, functionInstructions, outputDir, discovery.base);
  const calls = new Map<number, number>();
  for (const [source, target, mode] of discovery.calls) if (fn.instructions.has(source) && mode === "thumb") calls.set(source, target);
  const jumpTables = new Map<number, number[]>();
  for (const [site, table] of discovery.jump_table_sites) if (fn.instructions.has(site)) jumpTables.set(table, discovery.jump_tables.get(table)!);
  const labels = new Set<number>();
  for (const item of instructions) {
    const target = branchTarget(rom, item, discovery.instructions.get(item)!.kind, discovery.base);
    if (target !== null && fn.instructions.has(target)) labels.add(target);
  }
  for (const targets of jumpTables.values()) for (const target of targets) if (fn.instructions.has(target)) labels.add(target);
  const literals = new Map<number, number>();
  const lines = [".syntax unified", ".thumb", `glabel Func_${hex(address)}`];
  for (const item of instructions) {
    if (labels.has(item)) lines.push(`.L_${hex(item)}:`);
    let instruction = decoded.get(item)!.split(";", 1)[0].trim();
    const kind = discovery.instructions.get(item)!.kind;
    if (kind === "call" && calls.has(item)) instruction = `${instruction.split(/\s+/, 1)[0]} Func_${hex(calls.get(item)!)}`;
    else if (kind === "branch" || kind === "conditional") {
      const target = branchTarget(rom, item, kind, discovery.base)!;
      if (fn.instructions.has(target)) instruction = `${instruction.split(/\s+/, 1)[0]} .L_${hex(target)}`;
    }
    const half = discovery.u16(item);
    if ((half & 0xf800) === 0x4800) {
      const register = (half >>> 8) & 7;
      const literal = ((item + 4) & ~3) + ((half & 0xff) << 2);
      literals.set(literal, discovery.u32(literal));
      instruction = `ldr r${register}, .L_${hex(literal)}`;
    }
    lines.push(`    ${instruction}`);
  }
  for (const [literal, value] of [...literals].sort((a, b) => a[0] - b[0])) {
    const word = jumpTables.has(value) ? `.L_${hex(value)}` : `0x${hex(value)}`;
    lines.push("    .align 2", `.L_${hex(literal)}:`, `    .word ${word}`);
  }
  for (const [table, targets] of [...jumpTables].sort((a, b) => a[0] - b[0])) {
    lines.push("    .align 2", `.L_${hex(table)}:`, ...targets.map((target) => `    .word .L_${hex(target)}`));
  }
  writeFileSync(output, `${lines.join("\n")}\n`);
}
export const emit_discovery = emitDiscovery;

export function emit(rom: Uint8Array, address: number, output: string): void {
  const discovery = new Discovery(rom);
  discovery.run();
  emitDiscovery(rom, discovery, address, output);
}

function usage(): never {
  console.log("usage: emit_function.ts [-h] -o OUTPUT rom address");
  process.exit(0);
}

export function main(argv = process.argv.slice(2)): void {
  if (argv.includes("-h") || argv.includes("--help")) usage();
  let output: string | undefined;
  const positional: string[] = [];
  for (let index = 0; index < argv.length; index++) {
    if (argv[index] === "-o" || argv[index] === "--output") output = argv[++index];
    else positional.push(argv[index]);
  }
  if (positional.length !== 2 || !output) throw new Error("the following arguments are required: rom, address, -o/--output");
  emit(readFileSync(positional[0]), Number.parseInt(positional[1], 0) & ~1, output);
}

if (import.meta.main) main();
